new assumptions: use Z3 solver library to handle relational assumptions

TiloRC opened the pull request:
#### Brief description of what is fixed or changed

Use Z3 solver to add functionality so that the ask function can handle relational assumptions which it couldn’t handle at all for the most part. Queries to ask that involve a mix of relational assumptions and non-relational assumptions aren’t handled very well at the moment.

Example:

\>\>\> from sympy import ask
\>\>\> from sympy.abc import x, y, z
​​\>\>\> ask(x > z, (x > y) & (y > z))
True

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->


#### Other comments

This change breaks ask at the moment for some reason. I'm not totally sure why. 
ask(x > z, (x > y) & (y > z)) actually returns False for who knows what reason.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* assumptions
  * Added integration with Z3 solver so ask understands relational assumptions
<!-- END RELEASE NOTES -->

TiloRC committed:
new assumptions: use Z3 solver library to handle relational assumptions

Use Z3 solver to add functionality so that the ask function can handle
relational assumptions which it couldn’t handle at all for the most part.
Queries to ask that involve a mix of relational assumptions and
non-relational assumptions aren’t handled very well at the moment.

Example

>>> from sympy.abc import x, y, z
>>> ​​z3ask(x > z, (x > y) & (y > z))
True

moorepants commented on sympy/assumptions/tests/test_smtask.py:
```diff
@@ -0,0 +1 @@
    raises(ValueError, lambda:  z3ask((y > 0), (y > 2) & Q.imaginary(y)))
    assert z3ask((y > 0) & Q.imaginary(y), (y > 2))  is False # currently returns True
    assert z3ask(Q.real(y), y>2) is False
    assert z3ask( x > y, Q.infinite(x) & Q.positive(y) & Q.positive(x)) is True
```
It looks like you may not have git configured well for line endings. See https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

TiloRC committed:
fixed formatting

TiloRC committed:
Merge branch 'sympy:master' into add-smt-ask

asmeurer commented on sympy/assumptions/smtask.py:
```diff
@@ -0,0 +1,191 @@
+from sympy.logic.boolalg import And, Or, Not
+from sympy.assumptions.assume import global_assumptions, AppliedPredicate
+from sympy.printing.smtlib import smtlib_code
+import z3
```
We need to make z3 a proper optional dependency. See how this is done in other files (the use in `satisfiable` is a good model). It should also be added here https://github.com/sympy/sympy/blob/master/doc/src/contributing/dependencies.md

asmeurer commented on sympy/assumptions/smtask.py:
```diff
@@ -0,0 +1,191 @@
+
+    """
+
+    """
```
Use # to comment out code. Also ultimately commented out code should be removed, but it's OK if this is still a work-in-progress.

asmeurer commented on sympy/assumptions/smtask.py:
```diff
@@ -0,0 +1,191 @@
+    props = smtlib_code(proposition,auto_declare=False).replace("pow","^")
+    _props = smtlib_code(Not(proposition),auto_declare=False).replace("pow","^")
+
+    assumptions = smtlib_code(assumptions,auto_declare=False).replace("pow","^")
```
The pow replacement should be handled in the printer itself.

asmeurer commented on sympy/assumptions/tests/test_smtask.py:
```diff
@@ -0,0 +1,19 @@
+def test_z3ask_fail():
+    assert z3ask((y > 0), (y > 2) & Q.imaginary(y)) is False # currently returns True
+    raises(ValueError, lambda:  z3ask((y > 0), (y > 2) & Q.imaginary(y)))
+    assert z3ask((y > 0) & Q.imaginary(y), (y > 2))  is False # currently returns True
```
Let's also make sure we support the cases where Z3 can't handle the predicate at all. It should error rather than giving a wrong answer. If there aren't any such predicates, we can create a fake one.

TiloRC replied:
My changes to the SMT-Lib printer make it so that any predicate that's not handled is converted to a boolean dummy variable so there shouldn't be any need to throw an error anymore.

asmeurer commented on sympy/assumptions/smtask.py:
```diff
@@ -0,0 +1,191 @@
+        return formula
+
+
+def _pred_to_core(assmp):
```
It would be better to use the predicates as the canonical form, rather than the core Relational classes. One reason is that Relational classes have a lot of other logic on them which we will want to avoid, like automatic evaluation.

TiloRC committed:
Update to reflect changes in smtlib_code

I have been updating smtlib_code to handle predicate objects. I removed a lot
of bad code that assumed smtlib_code can’t handle predicate objects. I also
made some other minor improvements.

TiloRC committed:
Merge branch 'add-smt-ask' of https://github.com/TiloRC/sympy into add-smt-ask

TiloRC committed:
Make Z3 an optional dependency

TiloRC committed:
SMT-Lib printing: Handle predicate objects from new assumptions

The smtlib_code function will no longer throw an error when it tries to print some
predicate from the new assumptions. The printer throws an error if given a
predicate that can’t be translated to SMT-Lib or could be but hasn’t been
Implemented. While it’s possible to handle even, odd, and some other
predicate, this isn’t really all that important so this functionality was dropped.

Example:

In [1]: from sympy.printing.smtlib import smtlib_code

In [2]: from sympy.assumptions.ask import Q

In [3]:smtlib_code(Q.positive(x))

Out[3]]: '(declare-const x Real)\n(assert (> x 0))'

TiloRC committed:
Add test case for unhandled predicate

TiloRC committed:
Merge branch 'smtlib-predicate-simplified' into add-smt-ask

smt-ask is dependent on the code from smtlib-predicate.

TiloRC committed:
Add try catch to catch unhandled pred

TiloRC committed:
temporary fix: was using known_function para incorrectly

TiloRC commented:
I believe the approach being used in this pr is fundamentally wrong so I'm going to close it. I don't actually think any (or at least not much) of the code within the assumptions folder needs to be modified. 

Instead of writing an alternative function to satask, what I really should have been thinking about is writing an alternative to satisfiable from sympy.logic.inference. Or perhaps instead augmenting satisfiable to handle inequalities.

I may come back and reopen this PR if my new approach seems flawed for some reason.
